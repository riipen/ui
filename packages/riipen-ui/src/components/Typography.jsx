import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

const variantMapping = {
  h1: "h1",
  h2: "h2",
  h3: "h3",
  h4: "h4",
  h5: "h5",
  h6: "h6",
  subtitle1: "h6",
  subtitle2: "h6",
  body1: "p",
  body2: "p"
};

class Typography extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf([
      "initial",
      "inherit",
      "primary",
      "secondary",
      "tertiary",
      "positive",
      "negative"
    ]),

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     * By default, it maps the variant to a good default headline component.
     */
    component: PropTypes.elementType,

    /**
     * Controls the display type.
     */
    display: PropTypes.oneOf(["initial", "block", "inline"]),

    /**
     * If true, the text will have a bottom margin.
     */
    gutter: PropTypes.bool,

    /**
     * Set the text-align on the component.
     */
    textAlign: PropTypes.oneOf([
      "inherit",
      "left",
      "center",
      "right",
      "justify"
    ]),

    /**
     * Applies the theme typography styles.
     */
    variant: PropTypes.oneOf([
      "h1",
      "h2",
      "h3",
      "h4",
      "h5",
      "h6",
      "subtitle1",
      "subtitle2",
      "body1",
      "body2",
      "caption",
      "button",
      "inherit"
    ])
  };

  static defaultProps = {
    classes: [],
    color: "inherit",
    display: "initial",
    gutter: false,
    textAlign: "inherit",
    variant: "body1"
  };

  static contextType = ThemeContext;

  render() {
    const {
      children,
      classes,
      color,
      component,
      display,
      gutter,
      textAlign,
      variant
    } = this.props;

    const theme = this.context;

    const className = clsx(
      "root",
      `color-${color}`,
      display,
      gutter ? "gutter" : null,
      variant,
      `align-${textAlign}`,
      classes
    );

    const Component = component || variantMapping[variant] || "span";

    return (
      <React.Fragment>
        <Component className={className}>{children}</Component>
        <style jsx>{`
          .root {
            margin: 0;
          }

          .block {
            display: block;
          }
          .inline {
            display: inline;
          }

          .gutter {
            margin-bottom: 0.35em;
          }

          .color-inherit {
            color: inherit;
          }
          .color-initial {
            color: initial;
          }
          .color-primary {
            color: ${theme.palette.primary.main};
          }
          .color-secondary {
            color: ${theme.palette.secondary.main};
          }
          .color-tertiary {
            color: ${theme.palette.tertiary.main};
          }
          .color-positive {
            color: ${theme.palette.positive.main};
          }
          .color-negative {
            color: ${theme.palette.negative.main};
          }

          .align-center {
            text-align: center;
          }
          .align-left {
            text-align: left;
          }
          .align-right {
            text-align: right;
          }

          .body1 {
            font-family: ${theme.typography.body1.fontFamily};
            font-weight: ${theme.typography.body1.fontWeight};
            font-size: ${theme.typography.body1.fontSize};
            line-height: ${theme.typography.body1.lineHeight};
            letter-spacing: ${theme.typography.body1.letterSpacing};
          }
          .body2 {
            font-family: ${theme.typography.body2.fontFamily};
            font-weight: ${theme.typography.body2.fontWeight};
            font-size: ${theme.typography.body2.fontSize};
            line-height: ${theme.typography.body2.lineHeight};
            letter-spacing: ${theme.typography.body2.letterSpacing};
          }
          .button {
            font-family: ${theme.typography.button.fontFamily};
            font-weight: ${theme.typography.button.fontWeight};
            font-size: ${theme.typography.button.fontSize};
            line-height: ${theme.typography.button.lineHeight};
            letter-spacing: ${theme.typography.button.letterSpacing};
          }
          .caption {
            font-family: ${theme.typography.caption.fontFamily};
            font-weight: ${theme.typography.caption.fontWeight};
            font-size: ${theme.typography.caption.fontSize};
            line-height: ${theme.typography.caption.lineHeight};
            letter-spacing: ${theme.typography.caption.letterSpacing};
          }
          .h1 {
            font-family: ${theme.typography.h1.fontFamily};
            font-weight: ${theme.typography.h1.fontWeight};
            font-size: ${theme.typography.h1.fontSize};
            line-height: ${theme.typography.h1.lineHeight};
            letter-spacing: ${theme.typography.h1.letterSpacing};
          }
          .h2 {
            font-family: ${theme.typography.h2.fontFamily};
            font-weight: ${theme.typography.h2.fontWeight};
            font-size: ${theme.typography.h2.fontSize};
            line-height: ${theme.typography.h2.lineHeight};
            letter-spacing: ${theme.typography.h2.letterSpacing};
          }
          .h3 {
            font-family: ${theme.typography.h3.fontFamily};
            font-weight: ${theme.typography.h3.fontWeight};
            font-size: ${theme.typography.h3.fontSize};
            line-height: ${theme.typography.h3.lineHeight};
            letter-spacing: ${theme.typography.h3.letterSpacing};
          }
          .h4 {
            font-family: ${theme.typography.h4.fontFamily};
            font-weight: ${theme.typography.h4.fontWeight};
            font-size: ${theme.typography.h4.fontSize};
            line-height: ${theme.typography.h4.lineHeight};
            letter-spacing: ${theme.typography.h4.letterSpacing};
          }
          .h5 {
            font-family: ${theme.typography.h5.fontFamily};
            font-weight: ${theme.typography.h5.fontWeight};
            font-size: ${theme.typography.h5.fontSize};
            line-height: ${theme.typography.h5.lineHeight};
            letter-spacing: ${theme.typography.h5.letterSpacing};
          }
          .subtitle1 {
            font-family: ${theme.typography.subtitle1.fontFamily};
            font-weight: ${theme.typography.subtitle1.fontWeight};
            font-size: ${theme.typography.subtitle1.fontSize};
            line-height: ${theme.typography.subtitle1.lineHeight};
            letter-spacing: ${theme.typography.subtitle1.letterSpacing};
          }
          .subtitle2 {
            font-family: ${theme.typography.subtitle2.fontFamily};
            font-weight: ${theme.typography.subtitle2.fontWeight};
            font-size: ${theme.typography.subtitle2.fontSize};
            line-height: ${theme.typography.subtitle2.lineHeight};
            letter-spacing: ${theme.typography.subtitle2.letterSpacing};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Typography);
