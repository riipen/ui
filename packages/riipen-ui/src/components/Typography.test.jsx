import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Typography from "./Typography";

describe("<Typography>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Typography />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Typography />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Typography().type.defaultProps;

      const wrapper = mount(<Typography />);

      const component = wrapper.find("Typography");
      expect(component.props().color).toEqual(defaultProps.color);
      expect(component.props().display).toEqual(defaultProps.display);
      expect(component.props().gutter).toEqual(defaultProps.gutter);
      expect(component.props().textAlign).toEqual(defaultProps.textAlign);
      expect(component.props().variant).toEqual(defaultProps.variant);
    });
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <span>this is a child</span>;

      const wrapper = mount(<Typography>{child}</Typography>);

      expect(wrapper.find("Typography").contains(child)).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("applies classes to the root node of Typography", () => {
      const classes = ["classOne"];

      const wrapper = mount(<Typography classes={classes} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-typography"];

      const wrapper = mount(<Typography />);

      expect(
        wrapper
          .find("Typography")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("applies color class given a valid color", () => {
      const color = "white";

      const wrapper = mount(<Typography color={color} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(`color-${color}`)
      ).toEqual(true);
    });

    it("gives an error when given an invalid color", () => {
      const color = "red";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography color={color} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("component prop", () => {
    it("sets component with valid custom component", () => {
      const component = "a";

      const wrapper = mount(<Typography component={component} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .name()
      ).toEqual(component);
    });

    it("gives an error when given an invalid component", () => {
      const component = "invalidcomponent";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography component={component} />);

      expect(errors).toHaveBeenCalled();
    });

    it("applies the component prop over the variant prop when both are provided", () => {
      const component = "a";
      const variant = "h1";

      const wrapper = mount(
        <Typography component={component} variant={variant} />
      );

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .name()
      ).toEqual(component);
    });
  });

  describe("display prop", () => {
    it("sets display class with valid display", () => {
      const display = "inline";

      const wrapper = mount(<Typography display={display} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(display)
      ).toEqual(true);
    });

    it("gives an error when given an invalid display", () => {
      const display = "invalid";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography display={display} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("fontWeight prop", () => {
    it("sets fontWeight class with valid fontWeight", () => {
      const fontWeight = "medium";

      const wrapper = mount(<Typography fontWeight={fontWeight} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(fontWeight)
      ).toEqual(true);
    });

    it("gives an error when given an invalid fontWeight", () => {
      const fontWeight = "invalid";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography fontWeight={fontWeight} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("gutter prop", () => {
    it("applies gutter class when gutter prop is true", () => {
      const gutter = true;

      const wrapper = mount(<Typography gutter={gutter} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass("gutter")
      ).toEqual(true);
    });

    it("does not apply gutter class when gutter prop is false", () => {
      const gutter = false;

      const wrapper = mount(<Typography gutter={gutter} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass("gutter")
      ).toEqual(false);
    });
  });

  describe("textAlign prop", () => {
    it("sets textAlign class given a valid textAlign prop", () => {
      const textAlign = "justify";

      const wrapper = mount(<Typography textAlign={textAlign} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(`align-${textAlign}`)
      ).toEqual(true);
    });

    it("gives an error given an invalid textAlign prop", () => {
      const textAlign = "invalid";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography textAlign={textAlign} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("textTransform prop", () => {
    it("sets textTransform class given a valid textTransform prop", () => {
      const textTransform = "capitalize";

      const wrapper = mount(<Typography textTransform={textTransform} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(`transform-${textTransform}`)
      ).toEqual(true);
    });

    it("gives an error given an invalid textTransform prop", () => {
      const textTransform = "invalid";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography textTransform={textTransform} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("variant prop", () => {
    it("sets valid, mapped variant to component", () => {
      const variant = "h5";

      const wrapper = mount(<Typography variant={variant} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .name()
      ).toEqual(variant);
    });

    it("sets component to span when given an unmapped variant", () => {
      const variant = "inherit";

      const wrapper = mount(<Typography variant={variant} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .name()
      ).toEqual("span");
    });

    it("gives an error given an invalid variant", () => {
      const variant = "invalid";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography variant={variant} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("wordBreak prop", () => {
    it("sets wordBreak class given a valid wordBreak prop", () => {
      const wordBreak = "normal";

      const wrapper = mount(<Typography wordBreak={wordBreak} />);

      expect(
        wrapper
          .find("Typography")
          .childAt(0)
          .hasClass(`word-break-${wordBreak}`)
      ).toEqual(true);
    });

    it("gives an error given an invalid wordBreak prop", () => {
      const wordBreak = "invalid";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Typography wordBreak={wordBreak} />);

      expect(errors).toHaveBeenCalled();
    });
  });
});
