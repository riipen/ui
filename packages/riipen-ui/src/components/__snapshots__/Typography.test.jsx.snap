// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography> renders correct snapshot 1`] = `
<withClasses(Typography)>
  <Typography
    classes={
      Array [
        "riipen",
        "riipen-typography",
      ]
    }
    color="inherit"
    display="initial"
    gutter={false}
    textAlign="inherit"
    textTransform="inherit"
    variant="body1"
  >
    <p
      className="jsx-42457783 root color-inherit initial body1 align-inherit transform-inherit riipen riipen-typography"
    />
    <JSXStyle
      dynamic={
        Array [
          "#424242",
          "Roboto, Helvetica, Arial, sans-serif",
          400,
          "14px",
          "20px",
          "0.00938em",
          "#424242",
          "Roboto, Helvetica, Arial, sans-serif",
          400,
          "12px",
          "17px",
          "0.01071em",
          "#424242",
          "Roboto, Helvetica, Arial, sans-serif",
          400,
          "10px",
          "14px",
          "0.01071em",
          "#373737",
          "Roboto, Helvetica, Arial, sans-serif",
          500,
          "40px",
          "56px",
          "-0.01562em",
          "#373737",
          "Roboto, Helvetica, Arial, sans-serif",
          500,
          "30px",
          "42px",
          "-0.00833em",
          "#373737",
          "Roboto, Helvetica, Arial, sans-serif",
          500,
          "22px",
          "32px",
          "0em",
          "#373737",
          "Roboto, Helvetica, Arial, sans-serif",
          400,
          "20px",
          "30px",
          "-0.00735em",
          "#373737",
          "Roboto, Helvetica, Arial, sans-serif",
          400,
          "16px",
          "23px",
          "0rem",
          "#000",
          "#cccbcb",
          "#747474",
          "#424242",
          "#373737",
          "#1f71ae",
          "#f58632",
          "#3caabb",
          "#4cc88d",
          "#e95353",
          "#febb46",
          "#fff",
          700,
          300,
          500,
          400,
        ]
      }
      id="1840119567"
    />
  </Typography>
</withClasses(Typography)>
`;
